istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -0,0 +1,270 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/versioning"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth" // to avoid 'No Auth Provider found for name "gcp"'
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/clientcmd/api/latest"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/secretcontroller"
+)
+
+var (
+	Codec  runtime.Codec
+	Scheme *runtime.Scheme
+)
+
+func init() {
+	Scheme = runtime.NewScheme()
+	utilruntime.Must(v1.AddToScheme(Scheme))
+	yamlSerializer := json.NewYAMLSerializer(json.DefaultMetaFactory, Scheme, Scheme)
+	Codec = versioning.NewDefaultingCodecForScheme(
+		Scheme,
+		yamlSerializer,
+		yamlSerializer,
+		v1.SchemeGroupVersion,
+		runtime.InternalGroupVersioner,
+	)
+}
+
+type Options struct {
+	ServiceAccountName string
+	Contexts           []string
+
+	// inherited from root command
+	Namespace  string
+	Kubeconfig string
+}
+
+var (
+	defaultSecretPrefix       = "istio-pilot-remote-secret-"
+	DefaultServiceAccountName = "istio-pilot-service-account"
+)
+
+// NewCreatePilotRemoteSecretCommand creates a new command for joining two clusters togeather in a multi-cluster mesh.
+func NewCreatePilotRemoteSecretCommand(kubeconfig, namespace *string) *cobra.Command {
+	o := Options{
+		ServiceAccountName: DefaultServiceAccountName,
+	}
+
+	c := &cobra.Command{
+		Use:   "create-pilot-remote-secrets <cluster_context>... [OPTIONS]",
+		Short: "Create secret(s) for Pilot to discovery and join remote k8s service registries",
+		Example: `
+# Create a secret for remote cluster c1 and install it in cluster c0.
+istioctl x create-pilot-remote-secrets c1 > c1-secret.kubeconfigSecretYaml
+kubectl --context=c0 -n istio-system apply -f c1-secret.kubeconfigSecretYaml
+
+# Create and directly install secrets for multiple clusters. These 
+# commands 'join' clusters c0, c1, and c2 together.
+istioctl x create-pilot-remote-secrets c1 c2 \
+    | kubectl --context=c0 apply -f - --prune -l istio/multiCluster=true
+istioctl x create-pilot-remote-secrets c0 c2 \
+    | kubectl --context=c1 apply -f - --prune -l istio/multiCluster=true
+istioctl x create-pilot-remote-secrets c0 c1 \
+    | kubectl --context=c2 apply -f - --prune -l istio/multiCluster=true
+
```

Comment:
Not sure about this part. 
1. I'm a bit uncomfortable with keeping all those secrets on a local machine. We're talking about no-expiry bearer tokens here... Creating one and adding in one step ( with a istioctl crate | kubect apply ) seems safer.

2. I think for users it's simpler to just deal with 'add/remote' instead of attempting to use prune. Prune implies all secrets are included in bulk. Once a cluster is added - I would rather not have to mess with it again - keep in mind in some cases admin access to each cluster requires OTP, logins, etc.

Reply:
re: 1 - agreed that `create | apply` is preferable. We could also automatically `apply` the secret in the remote cluster, but we'd need to add more options to handle kubeconfig/context plus a corresponding `delete` command. Not sure if that's worth the trouble vs. relying on kubectl to do the heavy lifting.

re: 2 - I'm not sure if you're referring to the command usage or the command itself? The command can generate remote N secrets for N remote clusters output to the same file. This file can be applied to join a cluster to the N clusters. Are you suggesting to generate and apply/delete one secret at a time?